source/common/router/config_impl.cc

```diff
@@ -353,7 +358,37 @@ bool RouteEntryImplBase::evaluateRuntimeMatch(const uint64_t random_value) const
                                                        random_value);
 }
 
-bool RouteEntryImplBase::matchRoute(const Http::HeaderMap& headers, uint64_t random_value) const {
+bool RouteEntryImplBase::evaluateTlsContextMatch(const StreamInfo::StreamInfo& stream_info) const {
+  bool matches = true;
+
+  if (!tlsContextMatchCriteria()) {
+    return matches;
+  }
+
+  const TlsContextMatchCriteria& criteria = *tlsContextMatchCriteria();
+
+  if (criteria.presented().has_value()) {
+    const bool peerPresented = stream_info.downstreamSslConnection() &&
+                               stream_info.downstreamSslConnection()->peerCertificatePresented();
+    matches &= criteria.presented().value() == peerPresented;
+  }
+
+  if (matches && criteria.expired().has_value()) {
+    const SystemTime now = time_source_.systemTime();
+    const auto peerExpirationTime =
+        stream_info.downstreamSslConnection()->expirationPeerCertificate();
+    // Treats no presented credentials as not expired
+    const bool peerExpired =
+        peerExpirationTime.has_value() ? now > peerExpirationTime.value() : false;
```

Comment:
I figure this is a place that we can run into trouble; there should be one utility function throughout all of Envoy responsible for checking certification expiration and this should use that. The reason I am concerned is that different clock/time sources/methods might be used otherwise, and then we get inconsistent behaviors. Can you take a look at how this is done in the existing TLS context, RBAC etc. and at the implementation level refactor?

Reply:
I've not found other examples of certificate expiry checking. In the TLS context the permission to allow expired certs through is performed at the openssl callback level (checking the openssl error).
I could add to common/common/utility.h DateUtil class:
```
bool DateUtil::timePointExceeded(const SystemTime& time_point, TimeSource& time_source) {
  const auto now = time_source.systemTime();
  return now > time_point;
}
```
Then the code above becomes:
```
    const auto peer_expiration_time =
        stream_info.downstreamSslConnection()->expirationPeerCertificate();
    const bool peer_expired = peer_expiration_time.has_value() ? 
      DateUtil::timePointExceeded(peer_expiration_time.value(), time_source_) : false;
```
But I'm not sure about locking down the different clock/time sources.

Reply:
Why are we checking for expiration a second time? Can the TLS context let us know if the certificate was expired? I feel it is in the right place to make this determination. It's true that it might have expired since connection establishment, but is that a super important distinction to be made here? I feel with a reasonable assumption that connections are somewhat bounded in life (e.g. in the order of hours), it's fine to rely on the original expiration check. 

I definitely want to consult @PiotrSikora on this one, our resident SSL expert.

Reply:
Agreed - I'll look at pushing the expiry status through from the ssl verification callback.

Reply:
Should l remove the 'expired' route matching from this PR, and add it as another PR.
As we're trying to get PR #8248 through which adds checking whether a cert was 'validated', and adds the same 'plumbing' as is required for the 'expired' information through.
So this PR will primarily get the foundational API changes with example routing based on 'presented'

Reply:
I've removed the 'expired' TLSContext route matching from this PR to simplify to primarily the API changes that can be added to in subsequent PR's

Reply:
Thanks, this makes sense, let's followup on them in the future.